### ISSUE TITLE
Allow GUI to work without internet

### ISSUE THREAD

**Issue by TFWol** [oobabooga/text-generation-webui#867]
**Description**

If you clear the cache and block online parts like **fonts.googleapis.com**, the UI will be stuck at `Loading...` indefinitely.

Is it possible to make this more offline friendly?

### Edit
@Slins-23 Has a [great workaround](https://github.com/oobabooga/text-generation-webui/issues/867#issuecomment-1500288681) in the meantime.

**Comment by Slins-23** [comment]
I just lost my internet connection for half an hour and was wondering the same, so I glossed over the online requests, modified some lines in the Python scripts/HTML and was able to get it working without an internet connection.

The culprits were Google Fonts and Cloudflare's Iframe resize scripts, both of which are included by default on Gradio, at: `installer_files/env/lib/site-packages/gradio`.


Here's what I had to do:

1. Download iframeResizer source code from Cloudflare

Download `https://cdnjs.cloudflare.com/ajax/libs/iframe-resizer/4.3.6/iframeResizer.contentWindow.min.js` to `installer_files/env/lib/site-packages/gradio/templates/cdn/assets/iframeResizer.contentWindow.min.js`

Copy it to `installer_files/env/lib/site-packages/gradio/templates/frontend/assets/iframeResizer.contentWindow.min.js`

2. Delete Google Fonts requests and replace remote iframeResizer with local

At files `installer_files/env/lib/site-packages/gradio/templates/cdn/index.html`, `installer_files/env/lib/site-packages/gradio/templates/frontend/index.html`, and `installer_files/env/lib/site-packages/gradio/templates/frontend/share.html`, remove the <link> section that references `https://fonts.gstatic.com`, and replace `<script src="https://cdnjs.cloudflare.com/ajax/libs/iframe-resizer/4.3.1/iframeResizer.contentWindow.min.js"></script>` with `<script src="./assets/iframeResizer.contentWindow.min.js"></script>`.

Also remove `<script type="module" crossorigin src="https://gradio.s3-us-west-2.amazonaws.com/3.24.1/assets/index.18ea9dc6.js"></script>`, if it's in your file.

3. Fix Python files (there are quite a few, so watch out because it's error prone)

**THESE REFERENCE LINES HERE WILL CHANGE WHENEVER YOU ADD/REMOVE LINES, SO READ THE CODE TO SE IF IT MAKES SENSE, INSTEAD OF BLINDLY REPLACING/REMOVING THE LINE**

@`installer_files/env/lib/site-packages/gradio/test_data/blocks_configs.py`, line `190`, remove both `fonts.googleapis.com` strings. Do the same at line `476`.

@`installer_files/env/lib/site-packages/gradio/themes/__init__.py`, line `9`, in the `from gradio.themes.utils.fonts import Font, GoogleFont` expression, remove the `, GoogleFont` part.

@`installer_files/env/lib/site-packages/gradio/themes/base.py`, line `337` & `344`.

@`installer_files/env/lib/site-packages/gradio/themes/builder.py`

`204`: Replace `main_fonts, main_is_google = [], []` with `main_fonts = []`
`208`, `217`: Remove the 2 lines with `Google` in them.

`489`, `491`: Remove `+ main_is_google` and `+ mono_is_google`.

`536`, `537`: Remove both lines.

`544`, `545`: Remove `font_is_google` and `, pad_to_4(font_is_google)`, then do the same for `font_mono_is_google` below.

`566`, `568`: Remove both lines.

`667`: Remove the line `or isinstance(base_font, gr.themes.GoogleFont) != theme_font[1]`.

`671`: Replace with `f"'{font_name}'"`.

`764`: Delete `main_is_google` variable.

`782`: Delete `mono_is_google` variable.

`809`: Replace `for main_font, is_google in zip(main_fonts, main_is_google):` with `for main_font in main_fonts:`. Remove `if is_google:
                    main_font = gr.themes.GoogleFont(main_font, weights=font_weights)`

Do the same thing in the for loop below.

`856`, `857`: Replace with `list(main_fonts)`, `list(mono_fonts)`, respectively.

`871`: Replace the entire `_js` variable with `"""(css,fonts) => {}"""`.

`925`: Remove this for loop, which loops over `main_is_google + mono_is_google`.

@`installer_files/env/lib/site-packages/gradio/themes/default.py`

`22`, `30`: Delete these lines.

@`installer_files/env/lib/site-packages/gradio/themes/glass.py`

`31`: Remove this line.

@`installer_files/env/lib/site-packages/gradio/themes/monochrome.py`

`22`, `30`: Remove these lines.

@`installer_files/env/lib/site-packages/gradio/themes/soft.py`

`22`, `30`: Remove these lines.

@`installer_files/env/lib/site-packages/gradio/themes/utils/fonts.py`

`13`: Replace line with `"class": "font",`.

`22`: Replace line with `return Font(name)`.

`44`:  Delete the entire `GoogleFont` class.

**Comment by Slins-23** [comment]
Here's a gist that I made with all of these updated files: https://gist.github.com/Slins-23/9863726997a8933f46e5ee7d6eabc014

Each file with a "-" in the name indicates a nested folder, so i.e. "templates-cdn-index.html" == "templates/cdn/index.html".
This is assuming that the current working directory is at the gradio module root folder (`installer_files\env\lib\site-packages\gradio`)

**Comment by TFWol** [comment]
Sweet!
Thanks for the detailed response. I'll be diving into this as soon as I can get a bit of time.
This will be super helpful as a starting point when tackling other similar projects that use gradio.

Next will be figuring out how to stop gradio from bloody beaconing to api.gradio.app, despite the telemetry flags used.

**Comment by TFWol** [comment]
Works great! I went through the changes via WinMerge. That looks like it was very annoying to do.
Here's hoping @oobabooga will have this on radar.

**Comment by pm4user444** [comment]
I understand what it's doing, but the prior version didn't need to do this. Can you please alter this to not require internet once it's been installed?  Thanks

**Comment by TFWol** [comment]
@pm4user444 You might want to clarify your post a bit. It sounds like you're confusing us for the developer.
Maybe reword it to something like 'I have the same problem!'

**Comment by Slins-23** [comment]
@oobabooga You could use a custom version of gradio. You can fork that gradio version, make similar changes to the ones I did, and point to that repo instead of `gradio==3.24.1` at `text-generation-webui/requirements.txt`, by replacing it with `git+git://github.com/oobabooga/offline-gradio-fork.git`.

**Comment by oobabooga** [comment]
@Slins-23 you could submit your fixes as a PR to gradio

**Comment by Slins-23** [comment]
@oobabooga The problem is that they're more of a workaround than a fix. I didn't really come up with a sensible solution, I just deleted some references in a pretty nasty way, so it might have some ramifications that I don't know. As your repo uses a specific version, and it worked for me and @TFWol, there's at least some stability for this version specifically.

I just mentioned this issue in theirs, maybe they can look more specifically into these changes and workout a long-term solution. By the way, it seems that some previous versions didn't have this issue, so it might be the case that you could also downgrade.

**Comment by oobabooga** [comment]
I have disabled the use of Google Fonts in this commit, and the web UI seems to work offline now https://github.com/oobabooga/text-generation-webui/commit/649e4017a5646403f86c813a8471286cb5f98da2

@Slins-23 can you check?

**Comment by Slins-23** [comment]
@Slins-23 It does work fine for me now. Tested with [49aa05054ae13f381381440a9860ce0d68200e80 ](https://github.com/oobabooga/text-generation-webui/commit/649e4017a5646403f86c813a8471286cb5f98da2) and https://github.com/oobabooga/text-generation-webui/commit/49aa05054ae13f381381440a9860ce0d68200e80

**Comment by oobabooga** [comment]
Thanks for the confirmation. Since yesterday I have also blocked the HTTP request that gradio tries to make when it gets imported ([see here the fix](https://github.com/oobabooga/text-generation-webui/blob/main/server.py#L9)), so the only remote request left seems to be to

https://cdnjs.cloudflare.com/ajax/libs/iframe-resizer/4.3.1/iframeResizer.contentWindow.min.js

when the web UI loads on the client side. It should be possible to monkey patch this somehow by changing how this line behaves

https://github.com/gradio-app/gradio/blob/main/gradio/routes.py#L235

**Comment by Slins-23** [comment]
@oobabooga No problem, nice idea

**Comment by TFWol** [comment]
Confirmed working on my side too.
Thanks for making changes!

Should I do ahead and close out the issue?

**Comment by TFWol** [comment]
Confirmed working on my side too.
Thanks for making changes!

Should I go ahead and close out the issue?

**Comment by oobabooga** [comment]
Let's close this one since the original problem has been solved
